### PR TITLE
attempt to deal with waagent and docker racing

### LINKED ISSUE #3313: openshift: docker and waagent race condition for storage drive
On startup the waagent process and docker race to format disks:

From chat:
```
[12:36:46] <jminter> waagent and docker race at startup
[12:36:51] <jminter> docker starts
[12:37:09] <jminter> docker mounts sda2->/var/lib/docker/containers
[12:37:19] <jminter> waagent formats sdb1, mounts sdb1->/var/lib/docker
[12:37:31] <jminter> docker falls to pieces in a sorry heap
```

Mount command:
```
[root@ocp-master-24079797-0 ~]# mount
/dev/sda1 on / type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/sda2 on /var type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
/dev/sda2 on /var/lib/docker/containers type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
/dev/sda2 on /var/lib/docker/overlay2 type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
/dev/sdb1 on /var/lib/docker type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
```
docker-storage-setup file
```
[root@ocp-master-24079797-0 ~]# cat /etc/sysconfig/docker-storage-setup 
STORAGE_DRIVER=overlay2
```

Jim's first PR to tell waagent to not format the disk.
https://github.com/openshift/openshift-ansible/pull/8875

### PR DESCRIPTION
**Pull request by jim-minter**
fixes #3313

### CODE REVIEW

**Comment by jim-minter** [comment]
@pweil- @kwoodson (@kargakis) please can you look through this carefully

**Comment by jim-minter** [comment]
also see https://github.com/openshift/openshift-ansible/pull/8875 https://github.com/openshift/openshift-ansible/pull/8876

File: parts/openshift/release-3.9/openshiftmasterscript.sh [hunk 1]
```diff
@@ -31,13 +31,6 @@ else
 	COCKPIT_VERSION="latest"
 fi
 
-# TODO: with WALinuxAgent>=v2.2.21 (https://github.com/Azure/WALinuxAgent/pull/1005)
-# we should be able to append context=system_u:object_r:container_var_lib_t:s0
-# to ResourceDisk.MountOptions in /etc/waagent.conf and remove this stanza.
-systemctl stop docker.service
```
Comment (jim-minter) [inline parts/openshift/release-3.9/openshiftmasterscript.sh]: I believe the waagent changes to format and use /dev/sdb1 never made it into the 3.9 image so I believe this code was redundant?

File: parts/openshift/unstable/openshiftmasterscript.sh [hunk 2]
```diff
@@ -32,10 +32,33 @@ else
 	COCKPIT_VERSION="latest"
 fi
 
-# TODO: with WALinuxAgent>=v2.2.21 (https://github.com/Azure/WALinuxAgent/pull/1005)
-# we should be able to append context=system_u:object_r:container_var_lib_t:s0
-# to ResourceDisk.MountOptions in /etc/waagent.conf and remove this stanza.
+if grep -q ^ResourceDisk.Filesystem=xfs /etc/waagent.conf; then
```
Comment (jim-minter) [inline parts/openshift/unstable/openshiftmasterscript.sh]: think this is worthwhile to try to work around on the grounds that publishing a new image will take days

Comment (jim-minter) [reply parts/openshift/unstable/openshiftmasterscript.sh]: this codepath goes away eventually

File: parts/openshift/unstable/openshiftmasterscript.sh [hunk 3]
```diff
@@ -32,10 +32,33 @@ else
...
+	umount /var/lib/docker
+	umount /var/lib/docker/*
+
+	# disable waagent from racing again if we reboot.
+	sed -i -e '/^ResourceDisk.Format=/ s/=.*/=n/' /etc/waagent.conf
+	set -e
+fi
+
 systemctl stop docker.service
+mkfs.xfs -f /dev/sdb1
```
Comment (kwoodson) [inline parts/openshift/unstable/openshiftmasterscript.sh]: Is a format required?  I know waagent is no longer going to be performing the format.

Comment (jim-minter) [reply parts/openshift/unstable/openshiftmasterscript.sh]: yes - it comes in as ntfs, it needs to be xfs

File: parts/openshift/unstable/openshiftmasterscript.sh [hunk 4]
```diff
@@ -32,10 +32,33 @@ else
...
+if grep -q ^ResourceDisk.Filesystem=xfs /etc/waagent.conf; then
+	# Bad image: docker and waagent are racing.  Try to fix up.  Leave this code
+	# until the bad images have gone away.
+	set +e
+
+	# stop docker if it hasn't failed already
+	systemctl stop docker.service
+
+	# wait until waagent has run mkfs and mounted /var/lib/docker
+	while [[ $(stat -c %D /var/lib/docker) != 811 ]]; do
```
Comment (pweil-) [inline parts/openshift/unstable/openshiftmasterscript.sh]: where does 811 come from it doesn't seem static? When I boot a 3.10 cluster I'm seeing:
```
[root@ocp-master-25550318-0 ~]# stat -c %D /var/lib/docker
802
```
Is `mountpoint /var/lib/docker/containers` a good check for knowing it is mounted and `mount -l -t xfs` a better check to know we lost?  

```
[root@ocp-master-25550318-0 ~]# mountpoint /var/lib/docker/containers
/var/lib/docker/containers is a mountpoint
[root@ocp-master-25550318-0 ~]# mount -l -t xfs
/dev/sda1 on / type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/sda2 on /var type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
/dev/sda2 on /var/lib/docker/containers type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
/dev/sda2 on /var/lib/docker/overlay2 type xfs (rw,relatime,seclabel,attr2,inode64,grpquota)
```

Comment (pweil-) [reply parts/openshift/unstable/openshiftmasterscript.sh]: actually, maybe it's my input file on the device number but I guess the question is still valid.

Comment (jim-minter) [reply parts/openshift/unstable/openshiftmasterscript.sh]: `mountpoint /var/lib/docker/containers` also works fine, thanks - it's problematic when you know your system calls better than util-linux :/
updated.
I don't see where `mount -l -t xfs` would be needed / what it replaces?

**Comment by jim-minter** [comment]
@kargakis if you like it, please lgtm it :)

File: parts/openshift/unstable/openshiftmasterscript.sh (hunk 4 above)
Comment (pweil-) [reply parts/openshift/unstable/openshiftmasterscript.sh]: > I don't see where mount -l -t xfs would be needed / what it replaces?

was curious if we could tell from that call output if the race was lost or not.  Sounds like not.

**Pull request merged**